File: zig/1-blinky/perf.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const c = @import("c.zig");

pub const Marker = struct {
    time_ns: u64,

    pub fn now() Marker {
        var time: c.struct_timespec = undefined;
        const res = c.clock_gettime(c.CLOCK_THREAD_CPUTIME_ID, &time);
        const time_ns = if (res != 0) 0 else ns_from_timespec(&time);
        return Marker{ .time_ns = time_ns };
    }
};

pub const Counter = struct {
    const Self = @This();

    name: []const u8,
    samples_ns: std.ArrayList(u32),

    pub fn init(allocator: Allocator, name: []const u8, length: usize) !Self {
        var resolution: c.struct_timespec = undefined;
        const res = c.clock_getres(c.CLOCK_THREAD_CPUTIME_ID, &resolution);
        if (res != 0)
            return std.posix.unexpectedErrno(std.posix.errno(res));

        std.log.info(
            "Performance counter {s}, cpu resolution: {} ns",
            .{ name, ns_from_timespec(&resolution) },
        );

        const samples_ns = try std.ArrayList(u32).initCapacity(allocator, length);

        return .{
            .name = name,
            .samples_ns = samples_ns,
        };
    }

    pub fn deinit(self: *const Counter) void {
        self.samples_ns.deinit();
    }

    pub fn add_sample(self: *Self, start: Marker) void {
        const end = Marker.now();
        const diff = end.time_ns - start.time_ns;

        if (self.samples_ns.items.len >= self.samples_ns.capacity) {
            std.log.err("perf.Counter.add_sample: buffer is full", .{});
            return;
        }

        const sample = self.samples_ns.addOneAssumeCapacity();
        sample.* = @truncate(diff);
    }

    pub fn report(self: *const Self) void {
        std.log.info(
            "Performance counter {s}: {}",
            .{ self.name, SampleFormatter{ .data = self.samples_ns.items } },
        );
    }

    const SampleFormatter = std.fmt.Formatter(formatSamples);

    fn formatSamples(
        data: []const u32,
        comptime _: []const u8,
        _: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        try writer.writeAll("[");
        for (data, 0..) |sample, i| {
            try std.fmt.format(writer, "{d}", .{sample / 1000});
            if (i < data.len - 1)
                try writer.writeAll(",");
        }
        try writer.writeAll("] us");
    }

    pub fn reset(self: *Self) void {
        self.samples_ns.clearRetainingCapacity();
    }
};

fn ns_from_timespec(timespec: *const c.struct_timespec) u64 {
    const tv_sec: u64 = @intCast(timespec.tv_sec);
    const tv_nsec: u64 = @intCast(timespec.tv_nsec);
    return tv_nsec + tv_sec * std.time.ns_per_s;
}

File: c/1-blinky/perf.c
#include <bits/time.h>
#include <errno.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "perf.h"

uint64_t ns_from_timespec(const struct timespec *timespec) {
  return timespec->tv_nsec + timespec->tv_sec * 1000000000;
}

int perf_counter_init(
    perf_counter_t **const self, const char *name, size_t length
) {
  int res;

  struct timespec resolution;
  res = clock_getres(CLOCK_THREAD_CPUTIME_ID, &resolution);
  if (res != 0) {
    fprintf(stderr, "clock_getres fail (%d): %s\n", res, strerror(errno));
    return -1;
  }

  printf(
      "Performance counter %s, cpu resolution: %llu ns\n", name,
      ns_from_timespec(&resolution)
  );

  const size_t array_size = sizeof(uint32_t) * length;
  perf_counter_t *me = malloc(sizeof(perf_counter_t) + array_size);
  if (me == NULL) {
    fprintf(stderr, "malloc fail (%d): %s\n", errno, strerror(errno));
    return -1;
  }

  *me = (perf_counter_t){
      .name = name,
      .capacity = length,
      .length = 0,
  };

  *self = me;

  return 0;
}
void perf_counter_deinit(perf_counter_t *self) { free(self); }

perf_mark_t perf_mark() {
  struct timespec mark;
  int res = clock_gettime(CLOCK_THREAD_CPUTIME_ID, &mark);
  if (res != 0)
    return 0;
  else
    return ns_from_timespec(&mark);
}
void perf_counter_add_sample(perf_counter_t *self, perf_mark_t start) {
  const perf_mark_t end = perf_mark();
  const uint64_t diff = end - start;

  if (self->length >= self->capacity) {
    fprintf(stderr, "perf_counter_add_sample: buffer is full");
    return;
  }

  self->samples_ns[self->length] = diff;
  self->length += 1;
}

void perf_counter_report(perf_counter_t *const self) {
  printf("Performance counter %s: [", self->name);
  for (size_t i = 0; i < self->length; ++i) {
    printf("%u", self->samples_ns[i] / 1000);
    if (i < self->length - 1)
      printf(",");
  }
  printf("] us\n");
}
void perf_counter_reset(perf_counter_t *self) { self->length = 0; }
